pkg/dockerfile/fast_generator.go

```diff
@@ -362,3 +442,145 @@ func (g *FastGenerator) validateConfig() error {
 	}
 	return nil
 }
+
+func (g *FastGenerator) rsyncSrc(srcDir string, weights []weights.Weight) error {
+	matcher, err := dockerignore.CreateMatcher(g.Dir)
+	if err != nil {
+		return err
+	}
+
+	relPath, err := filepath.Rel(g.Dir, srcDir)
+	if err != nil {
+		return err
+	}
+
+	// Find files we haven't copied over yet.
+	usedFiles := make(map[string]bool)
+	usedFiles[relPath] = true
+	err = filepath.Walk(g.Dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if matcher != nil && matcher.MatchesPath(path) {
+			if info.IsDir() {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		if info.IsDir() && info.Name() == CogBuildArtifactsFolder {
+			return filepath.SkipDir
+		}
+
+		relPath, err := filepath.Rel(g.Dir, path)
+		if err != nil {
+			return err
+		}
+
+		// Skip weights, we handle them separately
+		for _, weight := range weights {
+			if weight.Path == relPath {
+				return nil
+			}
+		}
+
+		copyPath := filepath.Join(srcDir, relPath)
+		err = ensureFSObjectExists(copyPath, path)
+		if err != nil {
+			return err
+		}
+		usedFiles[relPath] = true
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Remove files that we no longer need in our tmp dir.
+	err = filepath.Walk(srcDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		relPath, err := filepath.Rel(srcDir, path)
+		if err != nil {
+			return err
+		}
+
+		_, ok := usedFiles[relPath]
+		if !ok {
+			console.Debug("Deleting " + relPath)
+			err = os.RemoveAll(path)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyFileWithTimes(destination string, src string) error {
+	console.Debug("Copying " + destination + " to " + src)
+	info, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+	err = files.CopyFile(src, destination)
+	if err != nil {
+		return err
+	}
+	err = os.Chtimes(destination, info.ModTime(), info.ModTime())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func ensureFSObjectExists(destination string, src string) error {
+	exists, err := files.Exists(destination)
+	if err != nil {
+		return err
+	}
+
+	info, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+	mode := info.Mode()
+
+	if !exists {
+		switch {
+		case mode.IsDir():
+			err := os.Mkdir(destination, 0o777)
+			if err != nil {
+				return err
+			}
+		case mode.IsRegular():
+			err := copyFileWithTimes(destination, src)
+			if err != nil {
+				return err
+			}
+		}
+	} else if mode.IsRegular() {
```

Comment:
Hard links work on Mac OS, so maybe it's easier than copying? IIRC they're guaranteed to have the same time, ownership etc. We just need to check src & dst have the same inode.

You might still wanna deal with symlinks though?

Reply:
Resolved in [5d27e08](https://github.com/replicate/cog/pull/2236/commits/5d27e0886afc2fceb29be1f2d5963905a1138e91)